Public/Src/Engine/Processes/SharedOpaqueOutputLogger.cs

```diff
@@ -0,0 +1,233 @@
+                }
+            }
+        }
+
+        /// <summary>
+        /// Same as <see cref="ReadRecordedPathsFromSharedOpaqueOutputLog(string)"/> except that all exceptions are wrapped in <see cref="BuildXLException"/>
+        /// </summary>
+        public static string[] ReadRecordedPathsFromSharedOpaqueOutputLogWrapExceptions(string filePath)
+        {
+            try
+            {
+                return ReadRecordedPathsFromSharedOpaqueOutputLog(filePath).ToArray();
+            }
+            catch (Exception e)
+            {
+                throw new BuildXLException("Failed to read from shared opaque output log", e);
+            }
+        }
+
+        /// <summary>
+        /// <see cref="RecordFileWrite(PathTable, AbsolutePath)"/>
+        /// </summary>
+        public bool RecordFileWrite(PathTable pathTable, string absolutePath)
+            => RecordFileWrite(pathTable, AbsolutePath.Create(pathTable, absolutePath));
+
+        /// <summary>
+        /// Records that the file at location <paramref name="path"/> was written to.
+        /// 
+        /// Returns whether the path was recorded or skipped.
+        /// </summary>
+        /// <returns>
+        /// <code>true</code> if <paramref name="path"/> was recorded, <code>false</code> 
+        /// if the path was filtered out because of <see cref="RootDirectories"/>.
+        /// </returns>
+        /// <remarks>
+        /// NOT THREAD-SAFE.
+        /// </remarks>
+        public bool RecordFileWrite(PathTable pathTable, AbsolutePath path)
+        {
+            if (RootDirectories == null || RootDirectories.Any(dir => path.IsWithin(pathTable, AbsolutePath.Create(pathTable, dir))))
+            {
+                if (m_recordedPathsCache.Add(path))
+                {
+                    m_lazyBxlWriter.Value.Write(path.ToString(pathTable));
+                    m_lazyBxlWriter.Value.Flush();
+                }
+
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        /// <nodoc />
+        public void Dispose()
+        {
+            m_lazyBxlWriter.Value.Dispose();
+        }
+
+        #region Serialization
+        /// <nodoc />
+        public void Serialize(BuildXLWriter writer)
+        {
+            writer.Write(LogPath);
+            writer.Write(RootDirectories, (w, list) => w.WriteReadOnlyList(list, (w2, path) => w2.Write(path)));
+        }
+
+        /// <nodoc />
+        public static SharedOpaqueOutputLogger Deserialize(BuildXLReader reader)
+        {
+            return new SharedOpaqueOutputLogger(
+                logPath: reader.ReadString(),
+                rootDirectories: reader.ReadNullable(r => r.ReadReadOnlyList(r2 => r2.ReadString())));
+        }
+        #endregion
+
+        private static string ReadStringOrNull(BuildXLReader bxlReader)
+        {
+            try
+            {
+                return bxlReader.ReadString();

```

Comment:
>ReadString [](start = 33, length = 10)

Should this be `ReadNullable<string>` ?

Reply:
no, this is `ReadString` or return `null` if failed to read string for any reason